librz/type/type.c

```diff
@@ -791,268 +791,305 @@ RZ_API ut64 rz_type_db_get_bitsize(const RzTypeDB *typedb, RZ_NONNULL RzType *ty
 	return 0;
 }
 
-struct HelperBufs {
-	RzStrBuf *arraybuf;
-	RzStrBuf *ptrbuf;
-};
+/**
+ * \brief Returns the type C representation
+ *
+ * \param typedb Types Database instance
+ * \param type RzType type
+ */
+RZ_API RZ_OWN char *rz_type_as_string(const RzTypeDB *typedb, RZ_NONNULL const RzType *type) {
+	rz_return_val_if_fail(typedb && type, NULL);
+
+	return rz_type_as_pretty_string(typedb, type, NULL, RZ_TYPE_PRINT_ZERO_VLA | RZ_TYPE_PRINT_NO_END_SEMICOLON | RZ_TYPE_PRINT_ANONYMOUS, 0);
+}
+
+/**
+ * \brief Returns the type C declaration representation
+ *
+ * \param typedb Types Database instance
+ * \param type RzType type
+ */
+RZ_API RZ_OWN char *rz_type_declaration_as_string(const RzTypeDB *typedb, RZ_NONNULL const RzType *type) {
+	rz_return_val_if_fail(typedb && type, NULL);
 
-static void helper_bufs_init(struct HelperBufs *hbs) {
-	hbs->arraybuf = rz_strbuf_new("");
-	hbs->ptrbuf = rz_strbuf_new("");
+	return rz_type_as_pretty_string(typedb, type, NULL, RZ_TYPE_PRINT_ZERO_VLA | RZ_TYPE_PRINT_NO_END_SEMICOLON, 1); // one level unfold
 }
 
-static void helper_bufs_fini(struct HelperBufs *hbs) {
-	rz_strbuf_free(hbs->arraybuf);
-	rz_strbuf_free(hbs->ptrbuf);
+/**
+ * \brief Returns the type C representation with identifier
+ *
+ * \param typedb Types Database instance
+ * \param type RzType type
+ */
+RZ_API RZ_OWN char *rz_type_identifier_declaration_as_string(const RzTypeDB *typedb, RZ_NONNULL const RzType *type, RZ_NONNULL const char *identifier) {
+	rz_return_val_if_fail(typedb && type, NULL);
+
+	return rz_type_as_pretty_string(typedb, type, identifier, RZ_TYPE_PRINT_ZERO_VLA | RZ_TYPE_PRINT_NO_END_SEMICOLON | RZ_TYPE_PRINT_UNFOLD_ANON_ONLY_STRICT, 1); // one level unfold (for anonymous only)
 }
 
-static char *type_as_string(const RzTypeDB *typedb, RZ_NONNULL const RzType *type, RZ_NONNULL struct HelperBufs *bufs) {
-	rz_return_val_if_fail(typedb && type && bufs, NULL);
+struct PrettyHelperBufs {
+	RzStrBuf *typename;
+	RzStrBuf *pointerbuf;
+	RzStrBuf *arraybuf;
+};
 
-	RzStrBuf *buf = rz_strbuf_new("");
+static bool type_decl_as_pretty_string(const RzTypeDB *typedb, const RzType *type, HtPP *used_types, struct PrettyHelperBufs phbuf, bool *self_ref, char **self_ref_typename, bool zero_vla, bool print_anon) {
+	rz_return_val_if_fail(typedb && type && used_types && self_ref, false);
+
+	bool is_anon = false;
 	switch (type->kind) {
 	case RZ_TYPE_KIND_IDENTIFIER: {
-		// Here it can be any of the RzBaseType
+		if (!type->identifier.name) {
+			return false;
+		}
+		is_anon = !strncmp(type->identifier.name, "anonymous ", 10);
```

Comment:
What is 10?
If it's the length of the string `anonymous `, just use `strlen("....")`, the compiler will optimize it and it's easier to understand.

Reply:
Thanks for pointing this out!

Also, I think I will make a different PR for refactoring all the `*types_as_c` functions. It will be easier to later bisect in case any of the refactoring break anything (also, since those functions don't seem to have tests, I do not want to introduce untested code (and probably bugs) in this PR).